File: lib/limousine/limousine_provider_eligibility.dart
import 'dart:math' as math;

import '../nearby/public_partner_bookability.dart';
import 'limousine_service_capability.dart';

/// Why a company is not a limousine provider. Missing evidence fails closed.
enum LimousineEligibilityDenial {
  missingCapability,
  companyInactive,
  companyDeleted,
  profileNotPublished,
  notBookable,
  noEligibleVehicle,
  marketMismatch,
}

class LimousineMarketRequest {
  const LimousineMarketRequest({
    this.postcode,
    this.countryCode,
    this.lat,
    this.lng,
  });

  final String? postcode;
  final String? countryCode;
  final double? lat;
  final double? lng;

  bool get hasConstraint {
    final pc = (postcode ?? '').trim();
    final cc = (countryCode ?? '').trim();
    return pc.isNotEmpty || cc.isNotEmpty || (lat != null && lng != null);
  }
}

class LimousineProviderEligibility {
  const LimousineProviderEligibility._({required this.eligible, this.denial});

  const LimousineProviderEligibility.allowed() : this._(eligible: true);

  const LimousineProviderEligibility.denied(LimousineEligibilityDenial denial)
    : this._(eligible: false, denial: denial);

  final bool eligible;
  final LimousineEligibilityDenial? denial;
}

/// Pure limousine provider gate.
///
/// A company is included only when every authoritative requirement is
/// present and consistent. Historical rides, premium tier, and a vehicle
/// named "limousine" never enable the company. No FLX company-code allowlist.
LimousineProviderEligibility evaluateLimousineProviderEligibility(
  Map<String, dynamic> company, {
  LimousineMarketRequest? request,
}) {
  if (_companyIsDeletedOrSuspended(company)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.companyDeleted,
    );
  }
  if (!_companyIsActive(company)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.companyInactive,
    );
  }
  if (!_publicProfilePublished(company)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.profileNotPublished,
    );
  }
  if (!isPublicPartnerBookable(company)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.notBookable,
    );
  }
  if (!partnerHasExplicitLimousineCapability(company)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.missingCapability,
    );
  }
  if (!_hasEligibleActiveLimousineVehicleOrService(company)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.noEligibleVehicle,
    );
  }
  if (request != null &&
      request.hasConstraint &&
      !_marketMatches(company, request)) {
    return const LimousineProviderEligibility.denied(
      LimousineEligibilityDenial.marketMismatch,
    );
  }
  return const LimousineProviderEligibility.allowed();
}

bool isEligibleLimousineProvider(
  Map<String, dynamic> company, {
  LimousineMarketRequest? request,
}) {
  return evaluateLimousineProviderEligibility(
    company,
    request: request,
  ).eligible;
}

List<Map<String, dynamic>> filterLimousineEligibleProviders(
  Iterable<Map<String, dynamic>> companies, {
  LimousineMarketRequest? request,
}) {
  return companies
      .where(
        (company) => isEligibleLimousineProvider(company, request: request),
      )
      .toList(growable: false);
}

bool isEligibleLimousineVehicle(Map<String, dynamic> vehicle) {
  if (_recordIsDeletedOrSuspended(vehicle)) return false;
  final hasExplicitLimousineConfig =
      // LIMOUSINE-MARKETPLACE-P2A: authoritative configured classification.
      isLimousineServiceToken(vehicle['service_category']?.toString()) ||
      isLimousineServiceToken(vehicle['serviceCategory']?.toString()) ||
      partnerHasExplicitLimousineCapability(vehicle) ||
      serviceCollectionContainsLimousine(vehicle['features']) ||
      serviceCollectionContainsLimousine(vehicle['service_ids']) ||
      serviceCollectionContainsLimousine(vehicle['serviceIds']) ||
      isLimousineServiceToken(vehicle['category']?.toString()) ||
      isLimousineServiceToken(vehicle['service']?.toString()) ||
      isLimousineServiceToken(vehicle['service_id']?.toString()) ||
      isLimousineServiceToken(vehicle['tierId']?.toString());
  if (!hasExplicitLimousineConfig) return false;
  // Published public vehicles often omit is_active; an explicit limousine
  // token may pass unless the row is explicitly inactive.
  return _recordIsActive(vehicle, missingMeansActive: true);
}

bool _hasEligibleActiveLimousineVehicleOrService(Map<String, dynamic> company) {
  if (_explicitServiceConfigurationsIncludeLimousine(company)) {
    return true;
  }
  final vehicles = _vehiclesFrom(company);
  if (vehicles == null) return false;
  for (final vehicle in vehicles) {
    if (isEligibleLimousineVehicle(vehicle)) return true;
  }
  return false;
}

bool _explicitServiceConfigurationsIncludeLimousine(
  Map<String, dynamic> company,
) {
  for (final key in const [
    'limousine_service_configurations',
    'limousineServiceConfigurations',
    'service_configurations',
    'serviceConfigurations',
  ]) {
    final raw = company[key];
    if (raw is! List) continue;
    for (final item in raw) {
      final row = asStringKeyedMap(item);
      if (row.isEmpty) continue;
      if (_recordIsDeletedOrSuspended(row)) continue;
      if (!_recordIsActive(row, missingMeansActive: true)) continue;
      if (partnerHasExplicitLimousineCapability(row) ||
          isLimousineServiceToken(row['service']?.toString()) ||
          isLimousineServiceToken(row['id']?.toString())) {
        return true;
      }
    }
  }
  return false;
}

List<Map<String, dynamic>>? _vehiclesFrom(Map<String, dynamic> company) {
  for (final key in const [
    'vehicles',
    'fleet',
    'public_vehicles',
    'publicVehicles',
  ]) {
    if (!company.containsKey(key)) continue;
    final raw = company[key];
    if (raw == null) return const <Map<String, dynamic>>[];
    if (raw is! List) return const <Map<String, dynamic>>[];
    return raw.whereType<Map>().map(asStringKeyedMap).toList(growable: false);
  }
  return null;
}

bool _companyIsDeletedOrSuspended(Map<String, dynamic> company) {
  if (_recordIsDeletedOrSuspended(company)) return true;
  final status = normalizePublicServiceToken(
    (company['status'] ?? company['company_status'] ?? company['companyStatus'])
        ?.toString(),
  );
  return status == 'deleted' ||
      status == 'suspended' ||
      status == 'tombstoned' ||
      status == 'tombstone';
}

bool _recordIsDeletedOrSuspended(Map<String, dynamic> record) {
  for (final key in const [
    'deleted',
    'is_deleted',
    'isDeleted',
    'tombstoned',
    'is_tombstoned',
    'isTombstoned',
    'suspended',
    'is_suspended',
    'isSuspended',
  ]) {
    if (looksTruthyPublicFlag(record[key])) return true;
  }
  return false;
}

bool _companyIsActive(Map<String, dynamic> company) {
  if (looksFalseyPublicFlag(company['is_active']) ||
      looksFalseyPublicFlag(company['isActive'])) {
    return false;
  }
  final availability = normalizePublicServiceToken(
    (company['availability_status'] ?? company['availabilityStatus'])
        ?.toString(),
  );
  if (availability == 'inactive' ||
      availability == 'deleted' ||
      availability == 'suspended') {
    return false;
  }
  if (looksTruthyPublicFlag(company['is_active']) ||
      looksTruthyPublicFlag(company['isActive'])) {
    return true;
  }
  if (availability == 'active') return true;
  return false;
}

bool _recordIsActive(
  Map<String, dynamic> record, {
  required bool missingMeansActive,
}) {
  if (looksFalseyPublicFlag(record['is_active']) ||
      looksFalseyPublicFlag(record['isActive'])) {
    return false;
  }
  if (looksTruthyPublicFlag(record['is_active']) ||
      looksTruthyPublicFlag(record['isActive'])) {
    return true;
  }
  return missingMeansActive;
}

bool _publicProfilePublished(Map<String, dynamic> company) {
  if (looksFalseyPublicFlag(company['profile_enabled']) ||
      looksFalseyPublicFlag(company['profileEnabled'])) {
    return false;
  }
  if (looksTruthyPublicFlag(company['profile_enabled']) ||
      looksTruthyPublicFlag(company['profileEnabled']) ||
      looksTruthyPublicFlag(company['public_profile_published']) ||
      looksTruthyPublicFlag(company['publicProfilePublished'])) {
    return true;
  }
  for (final key in const [
    'published_at',
    'publishedAt',
    'public_partner_profile_published_at',
    'publicPartnerProfilePublishedAt',
  ]) {
    if ((company[key] ?? '').toString().trim().isNotEmpty) return true;
  }
  return false;
}

bool _marketMatches(
  Map<String, dynamic> company,
  LimousineMarketRequest request,
) {
  final coverage = asStringKeyedMap(company['coverage']);
  final requestPostcode = _normalizePostcode(request.postcode);
  if (requestPostcode.isNotEmpty) {
    final companyPostcodes = <String>{
      _normalizePostcode(coverage['primary_postcode']?.toString()),
      _normalizePostcode(coverage['primaryPostcode']?.toString()),
      _normalizePostcode(company['primary_postcode']?.toString()),
      ..._postcodeList(coverage['postcodes']),
      ..._postcodeList(company['postcodes']),
    }..removeWhere((item) => item.isEmpty);
    if (companyPostcodes.isEmpty) return false;
    if (!companyPostcodes.contains(requestPostcode)) return false;
  }

  final requestCountry = _normalizeCountry(request.countryCode);
  if (requestCountry.isNotEmpty) {
    final companyCountry = _normalizeCountry(
      (coverage['country'] ??
              coverage['country_code'] ??
              coverage['countryCode'] ??
              company['country'] ??
              company['country_code'] ??
              company['countryCode'])
          ?.toString(),
    );
    if (companyCountry.isEmpty || companyCountry != requestCountry) {
      return false;
    }
  }

  if (request.lat != null && request.lng != null) {
    final lat = _asFiniteDouble(coverage['lat'] ?? company['lat']);
    final lng = _asFiniteDouble(coverage['lng'] ?? company['lng']);
    final radiusKm = _asFiniteDouble(
      coverage['service_radius_km'] ??
          coverage['serviceRadiusKm'] ??
          company['service_radius_km'] ??
          company['serviceRadiusKm'],
    );
    if (lat == null || lng == null || radiusKm == null || radiusKm <= 0) {
      return false;
    }
    if (_haversineKm(request.lat!, request.lng!, lat, lng) > radiusKm) {
      return false;
    }
  }
  return true;
}

String _normalizePostcode(String? raw) {
  return (raw ?? '').trim().toUpperCase().replaceAll(RegExp(r'\s+'), '');
}

String _normalizeCountry(String? raw) {
  return (raw ?? '').trim().toUpperCase();
}

Iterable<String> _postcodeList(dynamic raw) {
  if (raw is! List) return const <String>[];
  return raw.map((item) => _normalizePostcode(item?.toString()));
}

double? _asFiniteDouble(dynamic raw) {
  if (raw is num) return raw.toDouble();
  return double.tryParse((raw ?? '').toString().trim());
}

double _haversineKm(double lat1, double lng1, double lat2, double lng2) {
  const earthKm = 6371.0;
  final dLat = _degToRad(lat2 - lat1);
  final dLng = _degToRad(lng2 - lng1);
  final a =
      (math.sin(dLat / 2) * math.sin(dLat / 2)) +
      (math.cos(_degToRad(lat1)) *
          math.cos(_degToRad(lat2)) *
          math.sin(dLng / 2) *
          math.sin(dLng / 2));
  return earthKm * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
}

double _degToRad(double deg) => deg * math.pi / 180.0;

File: workers/booking/modules/limousine_provider_eligibility.mjs
// LIMOUSINE-MARKETPLACE-P1 — authoritative server-side provider eligibility.
//
// Pure, dependency-free module (mirrors the committed P0 Flutter contract in
// lib/limousine/*). It answers ONE question from authoritative projected state:
//
//   "May this company currently appear as a Limousine provider for this market?"
//
// Hard rules:
//   * fail closed on missing / stale / contradictory state;
//   * never infer from company name, vehicle name/brand/model, marketing words
//     (executive / premium / luxury / Mercedes), taxi tier, or historical use;
//   * entitlement (features['limousine']) is authoritative and server-owned;
//   * subscription lifecycle reuses the existing public-visibility semantics
//     (active / valid), never a new interpretation;
//   * older revisions may never overwrite a newer disable/suspension.

export const LIMOUSINE_SERVICE_TOKEN = "limousine";

const LIMOUSINE_SERVICE_ALIASES = new Set(["limousine", "limousine_service"]);

// Marketing/brand words that must NEVER, on their own, create eligibility.
const FORBIDDEN_CLASS_INFERENCE_TOKENS = new Set([
  "executive",
  "premium",
  "luxury",
  "vip",
  "business",
  "mercedes",
  "bmw",
  "audi",
  "tesla",
  "sclass",
  "s_class",
  "limo",
  "limousine",
  "comfort",
  "private",
]);

export const LIMOUSINE_ELIGIBILITY_REASONS = Object.freeze({
  ELIGIBLE: "eligible",
  COMPANY_INACTIVE: "company_inactive",
  COMPANY_SUSPENDED: "company_suspended",
  COMPANY_DELETED: "company_deleted",
  SUBSCRIPTION_NOT_PERMITTED: "subscription_not_permitted",
  NOT_ENTITLED: "not_entitled",
  NOT_ENABLED: "not_enabled",
  PROFILE_NOT_PUBLISHED: "profile_not_published",
  BOOKINGS_NOT_ACCEPTED: "bookings_not_accepted",
  MARKET_NOT_COVERED: "market_not_covered",
  NO_ELIGIBLE_VEHICLE: "no_eligible_vehicle",
});

export function normalizeLimousineToken(raw) {
  return String(raw ?? "")
    .trim()
    .toLowerCase()
    .replace(/[\s-]+/g, "_");
}

export function isLimousineServiceToken(raw) {
  return LIMOUSINE_SERVICE_ALIASES.has(normalizeLimousineToken(raw));
}

export function isForbiddenClassInferenceToken(raw) {
  return FORBIDDEN_CLASS_INFERENCE_TOKENS.has(normalizeLimousineToken(raw));
}

function looksTruthy(value) {
  if (value === true) return true;
  if (typeof value === "number") return value !== 0;
  const token = normalizeLimousineToken(value);
  return token === "true" || token === "1" || token === "yes" || token === "on" || token === "enabled";
}

function looksFalsey(value) {
  if (value === false) return true;
  if (typeof value === "number") return value === 0;
  const token = normalizeLimousineToken(value);
  return token === "false" || token === "0" || token === "no" || token === "off" || token === "disabled";
}

function asObject(raw) {
  return raw && typeof raw === "object" && !Array.isArray(raw) ? raw : {};
}

function serviceTokens(raw) {
  if (Array.isArray(raw)) {
    return raw.map((item) => normalizeLimousineToken(item)).filter((t) => t);
  }
  if (raw && typeof raw === "object") {
    return Object.keys(raw).map((k) => normalizeLimousineToken(k)).filter((t) => t);
  }
  return [];
}

/// Company explicitly enabled Limousine in its own public profile.
/// The durable opt-in is `services[]` containing a limousine token. A defaulted
/// `booking_capabilities.limousine=false` must not hide that token.
export function companyEnabledLimousine(candidate) {
  const c = asObject(candidate);
  if (serviceTokens(c.services).some((t) => LIMOUSINE_SERVICE_ALIASES.has(t))) {
    return true;
  }
  const capabilities = asObject(c.capabilities);
  const bookingCapabilities = asObject(c.booking_capabilities ?? c.bookingCapabilities);
  const explicit = [
    c.limousine_service_enabled,
    c.limousineServiceEnabled,
    capabilities.limousine,
    capabilities.limousine_service,
    bookingCapabilities.limousine,
    bookingCapabilities.limousine_service,
  ];
  let sawExplicit = false;
  let explicitTrue = false;
  let explicitFalse = false;
  for (const value of explicit) {
    if (value == null) continue;
    sawExplicit = true;
    if (looksTruthy(value)) explicitTrue = true;
    else if (looksFalsey(value)) explicitFalse = true;
  }
  if (explicitTrue) return true;
  if (explicitFalse || sawExplicit) return false;
  return false;
}

/// Authoritative, server-owned entitlement. Reads the projected boolean
/// (`limousine_entitled`) or the subscription features map. Missing / non-true
/// fails closed. Client-submitted values must already be stripped upstream.
export function resolveLimousineEntitlement(candidate) {
  const c = asObject(candidate);
  if (c.limousine_entitled === true) return true;
  if (c.limousine_entitled === false) return false;
  const features = asObject(c.features);
  if (features.limousine === true) return true;
  return false;
}

/// Public-discovery subscription gate. Reuses the existing
/// `_isSubscriptionActive` semantics (active / valid) — no new interpretation.
export function subscriptionPermitsLimousineDiscovery(status) {
  const s = normalizeLimousineToken(status);
  return s === "active" || s === "valid";
}

function recordIsBlocked(record) {
  const r = asObject(record);
  for (const key of [
    "deleted",
    "is_deleted",
    "isDeleted",
    "tombstoned",
    "is_tombstoned",
    "isTombstoned",
    "suspended",
    "is_suspended",
    "isSuspended",
  ]) {
    if (looksTruthy(r[key])) return true;
  }
  const status = normalizeLimousineToken(r.status ?? r.company_status ?? r.companyStatus);
  return status === "deleted" || status === "suspended" || status === "tombstoned" || status === "tombstone";
}

function recordIsActive(record, { missingMeansActive }) {
  const r = asObject(record);
  if (looksFalsey(r.is_active) || looksFalsey(r.isActive)) return false;
  if (looksTruthy(r.is_active) || looksTruthy(r.isActive)) return true;
  return missingMeansActive;
}

/// A vehicle/service qualifies only from authoritative configured identifiers:
/// service_category === 'limousine' AND a non-empty explicit service_class.
/// Never from name/brand/model/tier/marketing text.
export function isEligibleLimousineVehicle(vehicle) {
  const v = asObject(vehicle);
  if (recordIsBlocked(v)) return false;
  if (!recordIsActive(v, { missingMeansActive: true })) return false;
  const category = normalizeLimousineToken(v.service_category ?? v.serviceCategory);
  if (category !== LIMOUSINE_SERVICE_TOKEN) return false;
  const serviceClass = normalizeLimousineToken(v.service_class ?? v.serviceClass ?? v.service_class_id ?? v.serviceClassId);
  if (!serviceClass) return false;
  // A configured class id must not be a bare marketing/brand word.
  if (isForbiddenClassInferenceToken(serviceClass)) return false;
  return true;
}

function vehiclesFrom(candidate) {
  const c = asObject(candidate);
  const raw = c.vehicles ?? c.fleet ?? c.public_vehicles ?? c.publicVehicles;
  return Array.isArray(raw) ? raw : [];
}

export function hasEligibleLimousineVehicle(candidate) {
  return vehiclesFrom(candidate).some((v) => isEligibleLimousineVehicle(v));
}

function publicBookingsAccepted(candidate) {
  const c = asObject(candidate);
  if (looksFalsey(c.public_bookings_accepted) || looksFalsey(c.publicBookingsAccepted) || looksFalsey(c.bookable)) {
    return false;
  }
  const availability = normalizeLimousineToken(c.availability_status ?? c.availabilityStatus);
  if (availability === "inactive") return false;
  return true;
}

function profilePublished(candidate) {
  const c = asObject(candidate);
  if (looksFalsey(c.profile_enabled) || looksFalsey(c.profileEnabled)) return false;
  if (looksTruthy(c.profile_enabled) || looksTruthy(c.profileEnabled) || looksTruthy(c.public_profile_published)) {
    return true;
  }
  for (const key of ["published_at", "publishedAt", "public_partner_profile_published_at"]) {
    if (String(c[key] ?? "").trim()) return true;
  }
  return false;
}

function haversineKm(lat1, lng1, lat2, lng2) {
  const toRad = (d) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) * Math.sin(dLng / 2);
  return 6371 * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function normalizePostcode(raw) {
  return String(raw ?? "").trim().toUpperCase().replace(/\s+/g, "");
}

function marketCovered(candidate, request) {
  if (!request || typeof request !== "object") return true;
  const c = asObject(candidate);
  const coverage = asObject(c.coverage);
  const reqPostcode = normalizePostcode(request.postcode);
  if (reqPostcode) {
    const set = new Set(
      [
        normalizePostcode(coverage.primary_postcode ?? coverage.primaryPostcode),
        ...(Array.isArray(coverage.postcodes) ? coverage.postcodes.map(normalizePostcode) : []),
        ...(Array.isArray(c.supported_postcodes) ? c.supported_postcodes.map(normalizePostcode) : []),
      ].filter((x) => x),
    );
    if (set.size === 0 || !set.has(reqPostcode)) return false;
  }
  const reqCountry = normalizeLimousineToken(request.countryCode ?? request.country).toUpperCase();
  if (reqCountry) {
    const companyCountry = normalizeLimousineToken(
      coverage.country ?? coverage.country_code ?? c.country ?? c.country_code,
    ).toUpperCase();
    if (!companyCountry || companyCountry !== reqCountry) return false;
  }
  if (Number.isFinite(request.lat) && Number.isFinite(request.lng)) {
    const lat = Number(coverage.lat);
    const lng = Number(coverage.lng);
    const radiusKm = Number(coverage.service_radius_km ?? coverage.serviceRadiusKm);
    if (!Number.isFinite(lat) || !Number.isFinite(lng) || !Number.isFinite(radiusKm) || radiusKm <= 0) {
      return false;
    }
    if (haversineKm(request.lat, request.lng, lat, lng) > radiusKm) return false;
  }
  return true;
}

/// Authoritative eligibility composition. Returns `{ eligible, reason }`.
/// `reason` is a safe diagnostic code (no subscription-private detail).
export function evaluateLimousineProviderEligibility(candidate, { request = null } = {}) {
  const R = LIMOUSINE_ELIGIBILITY_REASONS;
  const c = asObject(candidate);

  if (recordIsBlocked(c)) return { eligible: false, reason: R.COMPANY_DELETED };
  if (!recordIsActive(c, { missingMeansActive: false })) {
    return { eligible: false, reason: R.COMPANY_INACTIVE };
  }
  if (!subscriptionPermitsLimousineDiscovery(c.subscription_status ?? c.subscriptionStatus)) {
    return { eligible: false, reason: R.SUBSCRIPTION_NOT_PERMITTED };
  }
  if (!resolveLimousineEntitlement(c)) return { eligible: false, reason: R.NOT_ENTITLED };
  if (!companyEnabledLimousine(c)) return { eligible: false, reason: R.NOT_ENABLED };
  if (!profilePublished(c)) return { eligible: false, reason: R.PROFILE_NOT_PUBLISHED };
  // Transaction gates, bookable flags and market/radius coverage never hide a
  // published limousine profile. `request` remains accepted for callers but is
  // ranking input only — nearby applies distance sort separately.
  if (request && typeof request === "object") {
    /* location is not a visibility filter */
  }
  if (!hasEligibleLimousineVehicle(c)) return { eligible: false, reason: R.NO_ELIGIBLE_VEHICLE };
  return { eligible: true, reason: R.ELIGIBLE };
}

export function isEligibleLimousineProvider(candidate, options = {}) {
  return evaluateLimousineProviderEligibility(candidate, options).eligible;
}

export function filterLimousineEligibleProviders(candidates, options = {}) {
  if (!Array.isArray(candidates)) return [];
  return candidates.filter((candidate) => isEligibleLimousineProvider(candidate, options));
}

/// Maps a `service` query param to a supported discovery filter. Unknown/empty
/// values return null (no service filter) and NEVER become limousine.
export function resolveNearbyServiceFilter(serviceParam) {
  const token = normalizeLimousineToken(serviceParam);
  if (!token) return null;
  if (LIMOUSINE_SERVICE_ALIASES.has(token)) return LIMOUSINE_SERVICE_TOKEN;
  return null;
}

/// Safe public capability signals to attach to a discovery payload. Contains
/// no subscription-private or customer-private data.
export function publicLimousineSignals(candidate) {
  return {
    limousine_available: isEligibleLimousineProvider(candidate),
    limousine_service_enabled: companyEnabledLimousine(candidate),
  };
}

/// Server-side entitlement projection stamped at publish time from the
/// authoritative subscription profile. Returns a single safe boolean.
export function projectLimousineEntitled({ features = null, subscriptionStatus = "" } = {}) {
  const f = asObject(features);
  const entitled = f.limousine === true;
  return entitled && subscriptionPermitsLimousineDiscovery(subscriptionStatus);
}

/// Monotonic availability transition: an older revision may never overwrite a
/// newer disable/suspension; equal revisions are idempotent replays; a newer
/// valid reactivation restores availability.
export function resolveLimousineAvailabilityTransition({
  currentCommand,
  currentRevision,
  incomingCommand,
  incomingRevision,
}) {
  const cur = Number(currentRevision) || 0;
  const next = Number(incomingRevision) || 0;
  if (next <= cur) {
    return {
      applied: false,
      effectiveCommand: currentCommand,
      effectiveRevision: cur,
      ignoredReason: next === cur ? "idempotent_replay" : "stale_revision",
    };
  }
  return {
    applied: true,
    effectiveCommand: incomingCommand,
    effectiveRevision: next,
    ignoredReason: null,
  };
}
